sdk/core/core-rest-pipeline/src/util/tokenCycler.ts

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AccessToken, GetTokenOptions, TokenCredential } from "@azure/core-auth";
+import { delay } from "./helpers";
+
+/**
+ * A function that gets a promise of an access token and allows providing
+ * options.
+ *
+ * @param options - the options to pass to the underlying token provider
+ */
+type AccessTokenGetter = (options: GetTokenOptions) => Promise<AccessToken>;
+
+export interface TokenCyclerOptions {
+  /**
+   * The window of time before token expiration during which the token will be
+   * considered unusable due to risk of the token expiring before sending the
+   * request.
+   *
+   * This will only become meaningful if the refresh fails for over
+   * (refreshWindow - forcedRefreshWindow) milliseconds.
+   */
+  forcedRefreshWindowInMs: number;
+  /**
+   * Interval in milliseconds to retry failed token refreshes.
+   */
+  retryIntervalInMs: number;
+  /**
+   * The window of time before token expiration during which
+   * we will attempt to refresh the token.
+   */
+  refreshWindowInMs: number;
+}
+
+// Default options for the cycler if none are provided
+export const DEFAULT_CYCLER_OPTIONS: TokenCyclerOptions = {
+  forcedRefreshWindowInMs: 1000, // Force waiting for a refresh 1s before the token expires
+  retryIntervalInMs: 3000, // Allow refresh attempts every 3s
+  refreshWindowInMs: 1000 * 60 * 2 // Start refreshing 2m before expiry
+};
+
+/**
+ * Converts an an unreliable access token getter (which may resolve with null)
+ * into an AccessTokenGetter by retrying the unreliable getter in a regular
+ * interval.
+ *
+ * @param getAccessToken - a function that produces a promise of an access
+ * token that may fail by returning null
+ * @param retryIntervalInMs - the time (in milliseconds) to wait between retry
+ * attempts
+ * @param timeoutInMs - the timestamp after which the refresh attempt will fail,
+ * throwing an exception
+ * @returns - a promise that, if it resolves, will resolve with an access token
+ */
+async function beginRefresh(
+  getAccessToken: () => Promise<AccessToken | null>,
+  retryIntervalInMs: number,
+  timeoutInMs: number
```

Comment:
nit: `timeoutInMs` really sounds like the number of ms to wait, not a date stamp in epoch ticks. Can we rename to something better? I'm struggling to come up with it though... /cc @bterlson any thoughts?

Reply:
What about `refreshExpiry`? @willmtemple 
I'm ok with any name, but I agree with Jeff.
We should also change this on the core-http copy.

Reply:
I’ll use refreshExpiry for now, Brian liked it over teams!

Reply:
Wait I like refreshTimeout more... I'll push refresh timeout for now